File: sorting_algorithms/src/mergeSort.cpp
#include <iostream>

using namespace std;

const int M = 5;
const int N = 10;


template <typename T> void printArray(T A[], int startIdx, int endIdx)
{
    cout << "Array: ";
    for (int i = startIdx; i <= endIdx; i++) { cout << A[i] << " "; }
    cout << endl;
}

template <typename T> T* createSubArray(T inputArray[], int startIdx, int endIdx)
{
    /* 
    Dynamically allocate new array so we can work with it outside this function
    ...remember to de-allocate with: 
        delete[] subarray;
    */
    int length = endIdx - startIdx + 1;
    int i = startIdx;// index for input array
    int j = 0;// index for sub array
    
    T* subarray = new T[length];
    
    while (i <= endIdx)
    {
        subarray[j] = inputArray[i];
        i++;
        j++;
    }
    
    return subarray;
}


template <typename T> void mergeSort(T A[], int startIdx, int endIdx)
{
    /* Generic merge sort */
    int length = endIdx - startIdx + 1;
    
    if (length > 1)
    {
        // Keep splitting array if we have more than one element
        int splitIdx = length / 2;
        
        T* L = createSubArray(A, startIdx, splitIdx-1); // copy left elements
        T* R = createSubArray(A, splitIdx, endIdx); // copy right elements
        
        mergeSort(L, 0, splitIdx-1-startIdx); // process left subarray
        mergeSort(R, 0, endIdx-splitIdx); // process right subarray
        
        // Merge...
        int lenL = splitIdx-1-startIdx+1;
        int lenR = endIdx-splitIdx+1;
        
        int i = 0;// track elements in L
        int j = 0;// track elements in R
        int k = 0;// track elements in A
        
        while (i < lenL && j < lenR)
        {
            // Insert smallest from either L or R until either one is entirely processed
            if (L[i] < R[j])
            {
                A[k] = L[i];
                i++;
            }
            else
            {
                A[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < lenL)
        {
            // Insert remaining elements from L, if any
            A[k] = L[i];
            i++;
            k++;
        }
        while (j < lenR)
        {
            // Insert remaining elements from R, if any
            A[k] = R[j];
            j++;
            k++;
        }
        
        // Cleanup
        delete[] L;
        delete[] R;
    }
}


int main()
{
    int A[M] = {7, 2, 8, 5, 4};
    int B[N] = {4, 5, 6, 8, 1, 2, 3, 5, 4, 7};
    
    printArray(A, 0, M-1);
    mergeSort(A, 0, M-1);
    printArray(A, 0, M-1);
    
    printArray(B, 0, N-1);
    mergeSort(B, 0, N-1);
    printArray(B, 0, N-1);
}

File: sorting_algorithms/src/merge_sort.py
#************************************************
'''
    Merge sort

    Any case run time: Theta(n*lg n)
'''
#************************************************
def merge_sort(A, verbose):
    if len(A) > 1:
        # Split array in half via integer/floor division
        mid = len(A)//2
        L = A[:mid]
        R = A[mid:]
        if verbose:
            print("A: ", A)
            print("Partitioning into:")
            print("  left: ", L)
            print("  right: ", R, "\n")

        # Keep splitting sub arrays until arrays are of size 1. This process takes Theta(lg n) time.
        merge_sort(L, verbose)
        merge_sort(R, verbose)

        if verbose:
            print("A: ", A)
            print("left array:  ", L)
            print("right array: ", R)
        
        # Initialize indices for...
        i = 0 # index to keep track of values in L
        j = 0 # index to keep track of values in R
        k = 0 # index to insert sorted values into A
        
        # Copy data from L[i] or R[j] to A[k].
        # See which value in L and R is smaller, insert that smaller element into A. This process takes Theta(n) time.
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                if verbose: 
                    print("inserting {} into index={} in A: {}".format(L[i], k, A))
                A[k] = L[i]
                i += 1
            else:
                if verbose: 
                    print("inserting {} into index={} in A: {}".format(R[j], k, A))
                A[k] = R[j]
                j += 1
            k += 1
        # Copy any remaining elements to A, obviously from smaller (L) to bigger (R)
        while i < len(L):
            A[k] = L[i]
            i += 1
            k += 1
        while j < len(R):
            A[k] = R[j]
            j += 1
            k += 1
    if verbose: 
        print("sorted array: ", A)
        if len(A) == 1:
            print("  recursion bottomed out")
        print("")

#================================================
# Sort some test arrays.
#================================================
A = [7, 2, 8, 5, 4]
B = [4, 5, 6, 8, 1, 2, 3, 5, 4, 7]
verbose = True

merge_sort(B, verbose)
